Admission Controller: Add Pod Preset

jessfraz opened the pull request:
Based off the proposal in https://github.com/kubernetes/community/pull/254
https://github.com/kubernetes/features/issues/162

cc @pmorie @pwittrock 

TODO:
- [x] tests

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**: Implements the Pod Injection Policy admission controller

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
Adds a new API resource `PodPreset` and admission controller to enable defining cross-cutting injection of Volumes and Environment into Pods.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41931)
<!-- Reviewable:end -->

pmorie commented on plugin/pkg/admission/podinjectionpolicy/admission.go:
```diff
@@ -0,0 +1,317 @@
+}
+
+// podInjectionPolicyPlugin is an implementation of admission.Interface.
+type podInjectionPolicyPlugin struct {
```
As we discussed, PSP needs to validate the policies here

pmorie commented on plugin/pkg/admission/podinjectionpolicy/admission.go:
```diff
@@ -0,0 +1,317 @@
+
+		glog.V(4).Infof("PodInjectionPolicy %s matches pod %s labels", pip.GetName(), pod.GetName())
+
+		// merge in policy for Env
```
Does this handle detecting conflicts?

pmorie replied:
TODO: add tests for this :)

pmorie commented on pkg/apis/apps/validation/validation.go:
```diff
@@ -129,3 +129,34 @@ func ValidateStatefulSetStatusUpdate(statefulSet, oldStatefulSet *apps.StatefulS
+// ValidatePodInjectionPolicySpec tests if required fields in the PodInjectionPolicy spec are set.
+func ValidatePodInjectionPolicySpec(spec *apps.PodInjectionPolicySpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
```
Validate that one of Env, EnvFrom, Volume, is set

pmorie replied:
Validate VolumeMount for each Volume

jessfraz committed:
Add settings API and admission controller

export functions from pkg/api/validation

add settings API

add settings to pkg/registry

add settings api to pkg/master/master.go

add admission control plugin for pod preset

add new admission control plugin to kube-apiserver

add settings to import_known_versions.go

add settings to codegen

add validation tests

add settings to client generation

add protobufs generation for settings api

update linted packages

add settings to testapi

add settings install to clientset

add start of e2e

add pod preset plugin to config-test.sh

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz commented:
I cannot figure out why this is happening on `make test-cmd` but I must have missed a place to add the new api group

```
k8s.io/kubernetes/pkg/controller/garbagecollector/garbagecollector.go:774: Failed to list <nil>: v1.ListOptions is not suitable for converting to "settings.k8s.io/v1alpha1"
```

smarterclayton commented:
Verify that `kube-controller-manager` has a dependency on `.../settings/install`.  Controller manager doesn't import pkg/master

jessfraz commented:
omg thank you, you might be my hero... we should probably document all
these (unless they are and i didnt see them) haha

On Mon, Feb 27, 2017 at 10:13 AM Clayton Coleman <notifications@github.com>
wrote:

> Verify that kube-controller-manager has a dependency on
> .../settings/install. Controller manager doesn't import pkg/master
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41931#issuecomment-282802206>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbJSRYEFyMPmUgDeJjjp91VyYpASYks5rgxJSgaJpZM4MJS-T>
> .
>

smarterclayton commented:
We document them, but we don't document them as fast as we add new things to document.

smarterclayton commented:
@sttts @deads2k can you guys make sure we have an issue to update the "adding an API group" based on everything in 1.6?

deads2k commented:
> @sttts @deads2k can you guys make sure we have an issue to update the "adding an API group" based on everything in 1.6?

@smarterclayton That one was you. :)  Opened https://github.com/kubernetes/kubernetes/issues/42180

I think in 1.7 we'll try to make the scheme and registry explicit so that its harder to miss.

MHBauer commented:
this is awesome! it does build on my machine (at least if it's tied into the default `make` invocation).

jessfraz commented:
yes it is tied in

pmorie reviewed (commented):
A couple nits, looking good so far.

jessfraz commented:
@k8s-bot test this please

pmorie commented:
LGTM - This is good enough for alpha and runs on Jess' local machine.  We likely need to find where to add admission controllers for some test contexts, but I don't think that should block this being alpha in 1.6

bgrant0607 commented:
@apelisse @grodrigues3 Does the bot respect manually applied "approved" labels?

bgrant0607 commented:
@pmorie @jessfraz 

I approve, but I believe my approval would be undone by the rebase that's needed. Ping me by slack or hangouts when this is ready for approval again.

bgrant0607 commented:
/lgtm
/approve

MHBauer commented:
Are the bots leading an uprising now?

bgrant0607 commented:
@jessfraz 

For future reference, squash guidelines:
https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md#8-fix-feedback-in-a-new-commit

Also, the release notes are for cluster administrators. We need better guidelines and examples of good release notes, but imagine someone reading your release note who has no idea what this is about.

duglin commented:
@jessfraz what level of docs are we planning on having for this for v1.6?

jessfraz commented:
I'll do docs after this is in, but yes it will be documented

On Tue, Feb 28, 2017, 08:31 Doug Davis <notifications@github.com> wrote:

> @jessfraz <https://github.com/jessfraz> what level of docs are we
> planning on having for this for v1.6?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41931#issuecomment-283090144>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbHF_vKQ3sARU-pdJdUfO1by41W4Qks5rhEvggaJpZM4MJS-T>
> .
>

grodrigues3 commented:
@bgrant0607 yes the bot respects human applied `approved` labels.  And if you type `/approve` it sticks despite rebases or when new code is committed (the exception being if you push code that requires approval from a previously un-needed owners file).

pwittrock commented:
@k8s-bot bazel test this

jessfraz committed:
run generators and updaters

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz committed:
update staging client

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz committed:
fix client-go

Signed-off-by: Jess Frazelle <acidburn@google.com>

jessfraz commented:
@k8s-bot non-cri e2e test this

jessfraz commented:
It's green!!!

pwittrock commented:
Adding P0 since this PR has been rebased several times already.

RenaudWasTaken commented:
Hi !

I've been trying to make this amazing PR work for my use case for a day now  :)

I can make it work as intended as long as my use case is at deployment time.
However as soon as my use case shifts to Bind time (SchedulerExtender) even
after tweaking it a bit, it seems to be rejected by the API-server.

To be a bit more specific I am trying to add an Env variable at Bind time which,
according to the issue abstract should be valid ?

>  Describes a policy resource that allows for the loose coupling of a Pod's definition from additional runtime requirements for that Pod. For example, mounting of Secrets, or setting additional environment variables, may not be known at Pod deployment time, but may be required at Pod creation time.

Do you think my use case is valid @jessfraz ? In which case is it a bug (and do you want me to open an issue) ?
Or is there a deeper rationale for not allowing updating Env variables at bind time ?

Thanks !

jessfraz commented:
I'm not sure what order of operations the Admission Controllers are running
in with regard to everything else. But I'd suspect that it's not working at
"bind time" (can explain what you are referring to here) because the
admission controller has not been called yet.

On Fri, Mar 24, 2017, 19:39 RenaudWasTaken <notifications@github.com> wrote:

> Hi !
>
> I've been trying to make this amazing PR work for my use case for a day
> now :)
>
> I can make it work as intended as long as my use case is at deployment
> time.
> However as soon as my use case shifts to Bind time (SchedulerExtender) even
> after tweaking it a bit, it seems to be rejected by the API-server.
>
> To be a bit more specific I am trying to add an Env variable at Bind time
> which,
> according to the issue abstract should be valid ?
>
> Describes a policy resource that allows for the loose coupling of a Pod's
> definition from additional runtime requirements for that Pod. For example,
> mounting of Secrets, or setting additional environment variables, may not
> be known at Pod deployment time, but may be required at Pod creation time.
>
> Do you think my use case is valid @jessfraz <https://github.com/jessfraz>
> ? In which case is it a bug (and do you want me to open an issue) ?
> Or is there a deeper rationale for not allowing updating Env variables at
> bind time ?
>
> Thanks !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41931#issuecomment-289179211>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbI7-TkDY4RO1NHd640xowS10UQjzks5rpHBQgaJpZM4MJS-T>
> .
>

RenaudWasTaken commented:
The "bind time" refers to "schedule time" when the scheduler assigns a node to the pod :)

Which means: 
    - it has already gone through the create operation in the admission controller
    - The Bind operation is an update (I've tweaked the AC to accept updates if the node wasn't assigned)

lavalamp commented on pkg/master/master.go:
```diff
@@ -371,6 +374,7 @@ func DefaultAPIResourceConfigSource() *serverstorage.ResourceConfig {
 		policyapiv1beta1.SchemeGroupVersion,
 		rbacv1beta1.SchemeGroupVersion,
 		rbacapi.SchemeGroupVersion,
+		settingsapi.SchemeGroupVersion,
```
Was there any discussion on this? We shouldn't be turning alpha APIs on by default.

@jessfraz @pwittrock

caesarxuchao replied:
/sub

jessfraz commented:
TBH I added it because I thought they all had to be there, so if this is
not the case and everything will compile without that, then it was not
intentional

On Thu, May 25, 2017 at 3:33 PM, Daniel Smith <notifications@github.com>
wrote:

> *@lavalamp* commented on this pull request.
> ------------------------------
>
> In pkg/master/master.go
> <https://github.com/kubernetes/kubernetes/pull/41931#discussion_r118566163>
> :
>
> > @@ -371,6 +374,7 @@ func DefaultAPIResourceConfigSource() *serverstorage.ResourceConfig {
>  		policyapiv1beta1.SchemeGroupVersion,
>  		rbacv1beta1.SchemeGroupVersion,
>  		rbacapi.SchemeGroupVersion,
> +		settingsapi.SchemeGroupVersion,
>
> Was there any discussion on this? We shouldn't be turning alpha APIs on by
> default.
>
> @jessfraz <https://github.com/jessfraz> @pwittrock
> <https://github.com/pwittrock>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41931#pullrequestreview-40371150>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABYNbI5ZraQWVebJZOgfBw1WCi37_cE0ks5r9deagaJpZM4MJS-T>
> .
>



-- 


Jessie Frazelle
4096R / D4C4 DD60 0D66 F65A 8EFC  511E 18F3 685C 0022 BFF3
pgp.mit.edu <http://pgp.mit.edu/pks/lookup?op=get&search=0x18F3685C0022BFF3>
